### PR TITLE
API Mocking Tools and Best Practices

### LINKED ISSUE #6226: [API]   API Mocking Tools and Best Practices
#### Proposed title of article
API Mocking Tools and Best Practices 

#### Proposed article introduction
The possibility to create mock-ups / mock system components is becoming increasingly important in contemporary development. It enables a number of development and testing workflows that greatly enhance the agility of developments teams in enterprises with complex infrastructures and systems. 

Just like testing, mocking can be done at many levels; code, API, service, and infrastructure to name some, and can be used for both functional and non-functional scenarios. Specifically, API mocking can empower your development and testing teams in their everyday work.

#### Key takeaways 

- An  overview of API mocking
- Real Backends vs Mock APIs
- Why API Mocking?
- API mocking best practices
- Examples of API mocking frameworks and tools
#### Article quality
This article digs deeper beyond explaining API mocking, unlike other available articles.  It compares mock APIs to real backends and explains the most popular and lesser-known API tools and frameworks while discussing their strengths and weaknesses.  

#### References
N/A

### PR DESCRIPTION
**Pull request by Polojigide**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [ ] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #6226.

### CODE REVIEW

**Comment by louisefindlay23** [comment]
Please ensure you link your topic suggestion to your article correctly in the future.

**Comment by louisefindlay23** [comment]
Plagiarism check complete

File: content/articles/api-mocking-tools-and-best-practices/index.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+Besides, testing the code that you write can be a little challenging. It entails testing the code that also performs HTTP requests. Making actual requests during the testing process may yield unreliable results. Mocking solves this issue by facilitating testing at different levels, including the code, API, service, and infrastructure. 
+
+This article looks into API mocking and compares mock APIs to real backends. It also discusses the most popular and lesser-known API tools and frameworks, including their strengths and weaknesses.
+
+### An overview of API mocking
+API mocking refers to imitations of the services and processes of a real API server to yield realistic mock API results for requests. Mock APIs recreate similar data representations as those displayed on data applications for real data. The objective is to simulate reality for precise assessment of the real-world applications of the software. Developers are keen to use API mocking to virtually recreate and assess the functioning of apps when deployed into the real world.
```
Comment (mbaabuones) [inline content/articles/api-mocking-tools-and-best-practices/index.md:10]: add content in this section.

File: content/articles/api-mocking-tools-and-best-practices/index.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+### How does API mocking work? 
+Some modifications are necessary to recreate a simulation of the interactions of a created application or software with users. New software that behaves like an original API replaces the authentic version of the API. But this new software lacks several functional and non-functional components of the real API. Thus, it can only simulate the results based on the requests but not produce the actual data picture.
```
Comment (mbaabuones) [inline content/articles/api-mocking-tools-and-best-practices/index.md:13]: add content in this section.

File: content/articles/api-mocking-tools-and-best-practices/index.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+#### Mockoon
+[Mockoon](https://mockoon.com/) is an API mocking tool. It has an efficient user interface and is built with Javascript and Electron.js. Of all the mock APIS,Mockoon is the easiest and fastest to run locally. It does not need remote deployment. However, it lacks features to visualize the input and output data at each endpoint. Besides, it cannot carry out proxy passes.
+
+#### Karate
+[Karate](https://github.com/intuit/karate/blob/master/karate-demo/src/test/java/mock/proxy/demo-mock.feature) is an open-source tool that combines mocks, performance testing, and API test automation into one framework. It supports native JavaScript Object Notation (JSON) and is easy to start. Also, it requires little coding. But this tool lacks IntelliSense support in IDE. It is difficult to debug the code, and it uses its own scripting language.
+
+### Moq
```
Comment (mbaabuones) [inline content/articles/api-mocking-tools-and-best-practices/index.md:58]: use level 4 header

File: content/articles/api-mocking-tools-and-best-practices/index.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+[Moq](https://github.com/moq/moq) is a mocking framework for C#/.NET. It has a good design and a consistent interface. But it does not perform ordered expectations and only mocks interfaces and abstract members.
+
+### SoapUI Mocks
```
Comment (mbaabuones) [inline content/articles/api-mocking-tools-and-best-practices/index.md:61]: use level 4 header

**Review by mbaabuones — CHANGES_REQUESTED** [review]
@Polojigide Great article! I have noted some grammatical errors in the article. Run your article through Grammarly to check and fix these errors. Please work on the minor changes suggested. Ping me when you are done.

**Comment by Polojigide** [comment]
I have edited this article @mbaabuones

**Review by mbaabuones — APPROVED** [review]

**Comment by hectorkambow** [comment]
jumping on this one

**Pull request merged**